pkg/compact/compact.go

```diff
@@ -278,34 +376,112 @@ func (c *Syncer) GarbageCollect(ctx context.Context) error {
 		}
+		c.metrics.garbageCollectedBlocks.Inc()
 	}
 	return nil
 }
 
+type groupMetrics struct {
+	cachedBlocksCount float64
+
+	compactions        prometheus.Counter
+	compactionFailures prometheus.Counter
+	compactionDuration prometheus.Histogram
+}
+
+func mapToString(m map[string]string) string {
+	var s string
+
+	var i int
+	for k, v := range m {
+		if i > 0 {
+			s += ";"
+		}
+		i++
+		s += fmt.Sprintf("%s=%s", k, v)
+	}
+	return s
+}
+
+func newGroupMetrics(reg *prometheus.Registry, labels map[string]string, getBlocksFunc func() (ids []ulid.ULID, err error)) *groupMetrics {
+	var m groupMetrics
+
+	// We cannot just append labels as metric labels, because it will conflict with the things generated by given Thanos
+	// source represent by this group.
+	groupLabels := map[string]string{
+		"group_labels": mapToString(labels),
+	}
+
+	groupBlocksCount := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name:        "thanos_compact_group_blocks_count",
+		Help:        "Number of currently loaded blocks for group.",
+		ConstLabels: groupLabels,
+	}, func() float64 {
+		b, err := getBlocksFunc()
+		if err != nil {
+			return float64(m.cachedBlocksCount)
+		}
+		m.cachedBlocksCount = float64(len(b))
+		return float64(len(b))
+	})
+
+	m.compactions = prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_compact_group_compactions_total",
+		Help:        "Total number of group compactions attempts.",
+		ConstLabels: groupLabels,
+	})
+	m.compactionFailures = prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_compact_group_compactions_failures_total",
+		Help:        "Total number of failed group compactions.",
+		ConstLabels: groupLabels,
+	})
+	m.compactionDuration = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:        "thanos_compact_group_compactions_duration_seconds",
+		Help:        "Time it took to compact whole group.",
+		ConstLabels: groupLabels,
+		Buckets: []float64{
+			0.25, 0.6, 1, 2, 3.5, 5, 7.5, 10, 15, 30, 60, 100, 200, 500,
+		},
+	})
+
+	if reg != nil {
+		reg.MustRegister(
+			groupBlocksCount,
+			m.compactions,
+			m.compactionFailures,
+			m.compactionDuration,
+		)
+	}
+	return &m
+}
+
 // Group captures a set of blocks that have the same origin labels.
 // Those blocks generally contain the same series and can thus efficiently be compacted.
 type Group struct {
-	logger log.Logger
-	mtx    sync.Mutex
-	dir    string
-	bkt    objstore.Bucket
-	labels map[string]string
+	logger  log.Logger
+	mtx     sync.Mutex
+	dir     string
+	bkt     objstore.Bucket
+	labels  map[string]string
+	metrics *groupMetrics
 }
 
-// NewGroup returns a new compaction group.
-func NewGroup(logger log.Logger, bkt objstore.Bucket, dir string, labels map[string]string) (*Group, error) {
+// newGroup returns a new compaction group.
+func newGroup(logger log.Logger, reg *prometheus.Registry, bkt objstore.Bucket, dir string, labels map[string]string) (*Group, error) {
```

Comment:
Why make it un-exported. The type is exported and from a package API perspective the only access to compaction functionality. It shouldn't be required to only be producible by the Sync.er

Reply:
No one wants it to be exported for now (: in terms of creation, right?

Reply:
Yea, but my take is that packages should provide some functionality (of course not unnecessarily generic) that allow them to be composed in different ways. 

The idea of only exporting what you use in your single use case got us to the point in Prometheus where virtually every package was some `type Manager struct` with two functions: `Manager.Run` and `Manager.Stop`. 
That's of course the extreme version.